README.md
[![Build Status](https://travis-ci.org/kanla/kanla.png?branch=master)](https://travis-ci.org/kanla/kanla)

You can find more information at http://kanla.zekjur.net/


Changes
Revision history for kanla:

1.5     Tue, 10 Jun 2014 20:13:37 +0200
        - actually read “interval” from config file
        - implement consecutive_failures option
        - implement silenced_by option
        - disable output buffering for better debugging
        - add “kanla-restarted” plugin to notify you when kanla crashed.
        - add and install kanla.service
        - http: use resolve_sockaddr so that “localhost” can come from /etc/hosts
        - properly encode UTF-8 in config parts that get passed to plugins
        - irc: react to ping messages

1.4     Fri, 25 Oct 2013 17:49:59 +0200
        - make Kanla::Plugin work reliably with newer Perl versions
	- add redis plugin
	- http plugin: implement basic auth
	- avoid using smartmatch since it is experimental in Perl ≥ 5.18

1.3     Thu, 23 May 2013 19:57:28 +0200
	- remove debug => 1 to remove errors when XML::Twig is missing
	- add "fail" plugin to the default config
	- install configfiles with permission 640
	- use kanla@example.invalid instead of example.com in the default config

1.2     Sat, 30 Mar 2013 12:26:46 +0100
	- Improve error message when having two <monitor> blocks with the same
	  name
	- Switch from MergeDuplicateOptions to manually adding defaults. This
	  fixes specifying multiple URLs.

1.1	Tue, 19 Mar 2013 15:39:00 +0100
        - Initial release.


Makefile.PL
use inc::Module::Install;
use Cwd;

name     'kanla';
all_from 'lib/Kanla.pm';
author   'Michael Stapelberg';

requires 'AnyEvent';
requires 'AnyEvent::Util';
requires 'AnyEvent::Handle';
requires 'AnyEvent::XMPP::Client';
requires 'AnyEvent::XMPP::Ext::Disco';
requires 'AnyEvent::XMPP::Ext::Ping';
requires 'AnyEvent::XMPP::Ext::VCard';
requires 'AnyEvent::XMPP::Ext::Version';
requires 'JSON::XS';
requires 'Config::General' => '2.50';
requires 'File::ShareDir';

# plugin requirements
requires 'AnyEvent::HTTP' => '1.5';
requires 'AnyEvent::Socket';
requires 'AnyEvent::DNS';

# test requirements
test_requires 'Test::MockModule';
test_requires 'Test::Deep';

install_script 'kanla';
# Install plugins/ into a distribution-shared directory
install_share('dist', 'plugins');

################################################################################
# We generate make targets for every file in docs/
# so users can use 'make docs'
# to generate docs/*.html from docs/*.txt
################################################################################

my @docs = <docs/*.txt>;
my $alldocfiles = join ' ', map { substr($_, 0, length() - 4) . '.html' } @docs;
my $postamble = '';
my $cwd = getcwd();

for my $file (@docs) {
	my $htmlfile = $file;
	$htmlfile =~ s/\.txt$/.html/;
$postamble .= <<EOT
$htmlfile: $file
	asciidoc -b html5 -f asciidoc.conf -a toc2 -a disable-javascript -a data-uri -a icons -a iconsdir=/etc/asciidoc/images/icons/ -a stylesheet=$cwd/asciidoc-toc.css -n -o - $file | hxtoc -l 2 > $htmlfile
EOT
}

postamble <<END_OF_MAKEFILE;
$postamble

docs: $alldocfiles
END_OF_MAKEFILE

postamble <<'END_OF_MAKEFILE';
install:: extra_install
pure_install:: extra_install
install_vendor:: extra_install

extra_install:
	install -d $(DESTDIR)/etc/kanla
	install -d $(DESTDIR)/etc/kanla/default.d
	install -d $(DESTDIR)/lib/systemd/system
	install -m 640 default.cfg $(DESTDIR)/etc/kanla/default.cfg
	install -m 640 default.d/* $(DESTDIR)/etc/kanla/default.d/
	install -m 644 kanla.service $(DESTDIR)/lib/systemd/system/kanla.service
END_OF_MAKEFILE

WriteAll;


lib/Kanla/Plugin.pm
# -*- Mode: CPerl;
# cperl-indent-level: 4;
# cperl-continued-statement-offset: 4;
# cperl-indent-parens-as-block: t;
# cperl-tabs-always-indent: t;
# cperl-indent-subs-specially: nil;
# -*-
# vim:ts=4:sw=4:expandtab
package Kanla::Plugin;

use strict;
use warnings;
use utf8;
use v5.10;

# libanyevent-perl
use AnyEvent;

# libconfig-general-perl
use Config::General;

# libjson-xs-perl
use JSON::XS;

# core
use IO::Handle;

use Exporter ();
our @EXPORT = qw(
    signal_error
    $conf
);

our $conf;

# see http://www.dagolden.com/index.php/369/version-numbers-should-be-boring/
our $VERSION = "1.5";
$VERSION = eval $VERSION;

# This will be set to a true value
# after initialization is complete,
# so that multiple callers can
# "use Kanla::Plugin"
# and all of them actually work :).
# (We do read STDIN for example,
#  which can only be done once.)
my $initialized;
my $errorfh;
my $main_timer;

sub signal_error {
    my ($severity, $message, $id) = @_;
    my $json = {
        severity => $severity,
        message  => $message
    };

    if (defined($id) && ref $id eq 'ARRAY') {
        $json->{id} = join('.', @$id);
    }

    say $errorfh encode_json($json);
}

sub import {
    my ($class, %args) = @_;

    say "kanla::plugin import";
    say "second!" if (defined($conf));

    my $pkg = caller;

    # Enable 5.10 features,
    # strict,
    # warnings,
    # utf8
    # for the caller.
    feature->import(":5.10");
    strict->import;
    warnings->import;
    utf8->import;

    # Setup the error fd
    if (!$initialized) {
        $errorfh = IO::Handle->new_from_fd(3, 'w');
        $errorfh->autoflush(1);

        # Parse the configuration
        my $config_str;
        {
            local $/;
            $config_str = <STDIN>;
        }

        do {
            $conf = Config::General->new(
                -String => $config_str,

                # open all files in utf-8 mode
                -UTF8 => 1,

                # normalize yes, on, 1, true and no, off, 0, false to 1 resp. 0
                -AutoTrue => 1,

                # case-insensitive key names by lowercasing everything
                -LowerCaseNames => 1,

                # provide the ->array, ->hash, etc. methods
                -ExtendedAccess => 1,

                -FlagBits => {
                    family => {
                        ipv4 => 1,
                        ipv6 => 1,
                    },
                },
            );

            if (!$conf->exists('family')) {
                $config_str .= <<'EOT';

family = ipv4 | ipv6
EOT
            }
        } until ($conf->exists('family'));

        my $interval =
            ($conf->exists('interval') ? $conf->value('interval') : 60);

        # Ensure timeout is an int and > 0.
        $interval += 0;
        $interval ||= 1;

        # Periodically run the check, but don’t wait for the first $interval seconds to
        # pass, but run it right now, too.
        my $run;
        $run = sub {
            # XXX ->can also resolves in @ISA
            die 'plugins must implement "sub run"'
                unless $run = $pkg->can("run");

            goto $run;
        };
        $main_timer = AnyEvent->timer(
            after    => 0,
            interval => $interval,
            cb       => $run,
        );

        $initialized = 1;
    }

    @_ = ($class);
    goto \&Exporter::import;
}

END {
    AnyEvent->condvar->recv;
}

1


lib/Kanla/Plugin/Banner.pm
# -*- Mode: CPerl;
# cperl-indent-level: 4;
# cperl-continued-statement-offset: 4;
# cperl-indent-parens-as-block: t;
# cperl-tabs-always-indent: t;
# cperl-indent-subs-specially: nil;
# -*-
# vim:ts=4:sw=4:expandtab
package Kanla::Plugin::Banner;

use strict;
use warnings;
use utf8;
use v5.10;

use Kanla::Plugin;

# libanyevent-perl
use AnyEvent;
use AnyEvent::Handle;
use AnyEvent::Socket;

use Socket qw(SOCK_STREAM);
use Exporter qw(import);
our @EXPORT = qw(
    banner_connect
    banner_disconnect
);

# see http://www.dagolden.com/index.php/369/version-numbers-should-be-boring/
our $VERSION = "1.5";
$VERSION = eval $VERSION;

# Filled in banner_connect().
my $timeout = 0;

=head1 NAME

Kanla::Plugin::Banner - Useful functions for banner-based plugins

=head1 SYNOPSIS

    use Kanla::Plugin;
    use Kanla::Plugin::Banner;

    sub run {
        banner_connect(
            host            => 'irc.twice-irc.de',
            default_service => 'ircd',
            cb              => sub {
                my ($handle, $timeout) = @_;
                $handle->push_write("NICK kanla\r\n");
                $handle->push_write("USER kanla kanla kanla :kanla\r\n");
                my @read_line;
                @read_line = (
                    line => sub {
                        my ($handle, $line) = @_;
                        if ($line !~ /^:[^ ]+ 001 /) {
                            $handle->push_read(@read_line);
                            return;
                        }

                        # We successfully signed on.
                        undef $timeout;
                        $handle->push_write("QUIT\r\n");
                        banner_disconnect($handle);
                    });

                $handle->push_read(@read_line);
            });
    }

=head1 METHODS

=cut

sub _banner_connect {
    my ($ip, $service, $cb) = @_;

    tcp_connect $ip, $service, sub {
        my ($fh) = @_;
        if (!$fh) {
            signal_error(
                'critical',
                "Connecting to $ip on port $service failed: $!"
            );
            return;
        }

        my $t;
        $t = AnyEvent->timer(
            after => $timeout,
            cb    => sub {
                signal_error(
                    'critical',

                    # XXX: It is unfortunate that this
                    # error message is so sparse.
                    # we should refactor the code to
                    # allow for better errors here.
                    "Timeout ($timeout s) on [$ip]:$service"
                );
            });

        my $handle;    # avoid direct assignment so on_eof has it in scope.
        $handle = AnyEvent::Handle->new(
            fh       => $fh,
            on_error => sub {
                signal_error(
                    'critical',
                    "TCP read error on [$ip]:$service: " . $_[2]);
                undef $t;
                $_[0]->destroy;
            },
            on_eof => sub {
                $handle->destroy;    # destroy handle
                signal_error(
                    'critical',
                    "TCP EOF on [$ip]:$service"
                );
                undef $t;
            });

        $cb->($handle, $t, $ip, $service);
    };
}

=head2 banner_connect

Connects to the given address
(parsed by C<AnyEvent::Socket>'s parse_hostport)
using the configured address families.

The caller provides a callback,
which will be called
after the connection was established.
In case there was an error
(DNS name could not be resolved,
connection was refused/timed out,
etc.),
the callback will B<NOT> be called,
but an alert will be signaled.

The callback will be called
with an C<AnyEvent::Handle> and
an C<AnyEvent> timer (timeouts).

The timeout is initialized
to the configured value
(plugin configuration)
or 20s if left unconfigured.

This example connects to one of
Google's SMTP servers
and waits for the SMTP greeting.
It does not resolve MX records,
but that's not the point of the example:

    banner_connect(
        host => 'aspmx.l.google.com',
        default_service => 'smtp',
        cb => sub {
            my ($handle, $timeout) = @_;
            $handle->push_read(line => sub {
                my ($handle, $line) = @_;
                undef $timeout;
                if ($line !~ /^220 /) {
                    signal_error('critical', 'Invalid greeting');
                }
            });
        });

=cut
sub banner_connect {
    my %args = @_;

    $timeout = ($conf->exists('timeout') ? $conf->value('timeout') : 20);

    # Ensure timeout is an int and > 0.
    $timeout += 0;
    $timeout ||= 1;

    my ($host, $service) =
        parse_hostport($args{'host'}, $args{'default_service'});

    my $resolved_cb = sub {

        # family is either A or AAAA
        my $family = shift;
        if (@_ == 0) {
            signal_error(
                'critical',
                "Cannot resolve $args{'host'} ($family) DNS record"
            );
            return;
        }
        for my $record (@_) {
            my ($service, $host) =
                AnyEvent::Socket::unpack_sockaddr($record->[3]);
            _banner_connect(format_address($host), $service, $args{'cb'});
        }
    };

    if ($conf->obj('family')->value('ipv4')) {
        AnyEvent::Socket::resolve_sockaddr(
            $host, $service, "tcp", 4, SOCK_STREAM,
            sub { $resolved_cb->('A', @_) });
    }
    if ($conf->obj('family')->value('ipv6')) {
        AnyEvent::Socket::resolve_sockaddr(
            $host, $service, "tcp", 6, SOCK_STREAM,
            sub { $resolved_cb->('AAAA', @_) });
    }
}

=head2 banner_disconnect($handle)

Properly disconnects
the specified C<AnyEvent::Handle>.

=cut
sub banner_disconnect {
    my ($handle) = @_;

    $handle->on_drain(
        sub {
            shutdown $handle->{fh}, 1;
            $handle->destroy;
            undef $handle;
        });
}

1


lib/Kanla.pm
# -*- Mode: CPerl;
# cperl-indent-level: 4;
# cperl-continued-statement-offset: 4;
# cperl-indent-parens-as-block: t;
# cperl-tabs-always-indent: t;
# cperl-indent-subs-specially: nil;
# -*-
package Kanla;

use strict;
use warnings;
use utf8;
use v5.10;

# libanyevent-xmpp-perl
use AnyEvent::XMPP::Client;
use AnyEvent::XMPP::Ext::Disco;
use AnyEvent::XMPP::Ext::Ping;
use AnyEvent::XMPP::Ext::VCard;
use AnyEvent::XMPP::Ext::Version;
# ::Receipts was added in AnyEvent::XMPP 0.54.
# kanla works fine without,
# but it’s nice to have.
# Therefore, load it and fall back if that fails.
my $use_receipts = eval {
    require AnyEvent::XMPP::Ext::Receipts;
    AnyEvent::XMPP::Ext::Receipts->import();
    1;
};

# libfile-sharedir-perl
use File::ShareDir qw(dist_dir);

# libanyevent-perl
use AnyEvent;
use AnyEvent::Util;
use AnyEvent::Handle;

# libconfig-general-perl
use Config::General;

# libjson-xs-perl
use JSON::XS;

# core
use Cwd qw(abs_path);
use Carp;
use Data::Dumper;
use File::Basename qw(basename);
use Encode qw(encode decode);

# see http://www.dagolden.com/index.php/369/version-numbers-should-be-boring/
our $VERSION = "1.5";
$VERSION = eval $VERSION;

binmode STDOUT, ':utf8';
binmode STDERR, ':utf8';

my $conf;
my $xmpp;

# Messages which were produced while no XMPP connection was established (yet).
# They will be sent when a connection is established.
my @queued_messages;

# Returns the path
# to the specified plugin
# by first checking dist_dir('kanla'),
# then 'plugins/',
# just like the configuration
# is searched in /etc/kanla,
# then in '.'.
sub plugin_path {
    my ($plugin) = @_;

    # The custom plugin dir
    # takes precedence.
    for my $dir (qw(share lib)) {
        my $path = "/usr/local/$dir/kanla/$plugin";
        return $path if -e $path;
    }

    my $dist_dir;
    # We eval because
    # dist_dir dies
    # when the dir does not exist.
    eval { $dist_dir = dist_dir('kanla'); };
    if (defined($dist_dir)) {
        return "$dist_dir/$plugin" if -e "$dist_dir/$plugin";
    }

    # NB: This does not imply
    # that the plugin exists.
    return "plugins/$plugin";
}

sub start_plugin {
    my ($plugin, $name) = @_;

    # Save the config for this plugin to string,
    # we will feed it to the plugin via stdin below.
    my $config = $conf->obj('monitor')->obj($name);
    # We need to specifically encode the config string
    # because AnyEvent does not
    # specify an encoding.
    my $config_str = encode('UTF-8', $config->save_string());

    my @dest;
    if ($config->exists('send_alerts_to')) {
        @dest = split("\n", $config->value('send_alerts_to'));
    } else {
        @dest = split("\n", $conf->value('send_alerts_to'));
    }

    say qq|[$plugin/instance "$name"] starting…|;

    my ($pr, $pw) = AnyEvent::Util::portable_pipe;
    fcntl($pr, AnyEvent::F_SETFD, AnyEvent::FD_CLOEXEC);
    my $w;
    $w = AnyEvent::Handle->new(
        fh       => $pr,
        on_error => sub {
            my ($hdl, $fatal, $msg) = @_;
            say STDERR
                qq|[$plugin/instance "$name"] error reading from stderr: $msg|;

            # Restart the plugin,
            # so that you can just kill plugins
            # after changing their code.
            #
            # The delay of 2 seconds avoids
            # spamming the user with errors
            # when a plugin exits immediately.
            my $t;
            $t = AnyEvent->timer(
                after => 2,
                cb    => sub {
                    start_plugin($plugin, $name);
                    undef $t;
                });
            $w->destroy;
        });

    my @start_request;
    @start_request = (
        json => sub {
            my ($hdl, $hashref) = @_;
            handle_stderr_msg($name, basename($plugin), \@dest, $hashref);
            $hdl->push_read(@start_request);
        });

    $w->push_read(@start_request);

    my $cv = run_cmd [ plugin_path($plugin) ],

        # feed the config on stdin
        '<', \$config_str,

        # stdout goes to /dev/null for now.
        '>', '/dev/null',

        # TODO: proxy stderr into our log so that one can easily spot plugin failures
        '3>', $pw;
    $cv->cb(
        sub {
            my $status = shift->recv;
            say STDERR
                qq|[$plugin/instance "$name"] exited with exit code $status|;
        });
}

sub message_silenced {
    my ($message, $counts) = @_;

    my $pattern = $message->{silenced_by};
    if (!defined($pattern) &&
        $conf->exists('silenced_by')) {
        $pattern = $conf->value('silenced_by');
    }

    return unless defined($pattern);

    my @matches;
    my @ids = grep { $_ ne $message->{failure_id} } keys %$counts;
    if (my ($re) = ($pattern =~ m,^/(.*)/$,)) {
        # Regular expression matching
        my $compiled = qr/$re/;
        @matches = grep { /$compiled/ } @ids;
    } else {
        # String prefix matching
        my $prefix = quotemeta($pattern);
        @matches = grep { m,^$prefix, } @ids;
    }

    return @matches > 0;
}

sub xmpp_empty_queue {
    my @leftover;

    my $consecutive_failures = (
        $conf->exists('consecutive_failures') ?
            $conf->value('consecutive_failures') :
            0
    );

    my %counts;
    for my $entry (@queued_messages) {
        my ($jid, $message) = @$entry;
        $counts{ $message->{failure_id} }++;
    }

    for my $entry (@queued_messages) {
        my ($jid, $message) = @$entry;

        next if message_silenced($message, \%counts);

        if ($counts{ $message->{failure_id} } < $consecutive_failures) {
            push @leftover, $entry unless $message->{expiration} < time();
            next;
        }

        # Find the account,
        # if unsuccessful,
        # no account
        # is connected.
        my $account = $xmpp->find_account_for_dest_jid($jid);
        if (!defined($account) || !defined($account->connection)) {
            push @leftover, $entry;
            next;
        }

        my $presence = $xmpp->get_priority_presence_for_jid($jid);
        if (!defined($presence)) {
            say "[XMPP] No presence found for $jid, skipping";
            push @leftover, $entry;
            next;
        }

        # NB: We cannot use $xmpp->send_message here because
        # that will make the JID a bare JID and use its own
        # conversation tracking technique.
        $account->connection->send_message(
            $presence->jid,
            'chat',
            undef,
            body => $message->{body});
    }

    @queued_messages = @leftover;
}

sub handle_stderr_msg {
    my ($name, $module, $dest, $data) = @_;
    if (!exists($data->{severity}) ||
        !exists($data->{message})) {
        say STDERR
"Malformed JSON output from module $module (missing severity or messages property).";
        return;
    }

    my $module_config = $conf->obj('monitor')->obj($name);
    my $interval      = 60;
    if ($module_config->exists('interval')) {
        $interval = $module_config->value('interval');
    }
    my $silenced_by = undef;
    if ($module_config->exists('silenced_by')) {
        $silenced_by = $module_config->value('silenced_by');
    }

    if ($data->{severity} eq 'critical') {
        say "read from plugin: " . $data->{message};
        my $message = {
            body        => $data->{message},
            expiration  => time() + $interval,
            failure_id  => $data->{id} // $data->{message},
            silenced_by => $silenced_by,
        };

        for my $jid (@$dest) {
            push @queued_messages, [ $jid, $message ];
        }
        xmpp_empty_queue();
    }
}

sub run {
    my %args = @_;

    $args{configfile} //= 'default.cfg';

    $conf = Config::General->new(

        # XXX: Not sure if '.' is a good idea. It makes development easier.
        -ConfigPath => [ '/etc/kanla', '.' ],
        -ConfigFile => $args{configfile},

        # open all files in utf-8 mode
        -UTF8 => 1,

        # normalize yes, on, 1, true and no, off, 0, false to 1 resp. 0
        -AutoTrue => 1,

        # case-insensitive key names by lowercasing everything
        -LowerCaseNames => 1,

        # include files relative to the location
        -IncludeRelative => 1,

        # allow glob patterns in include statements
        -IncludeGlob => 1,

        # allow including the same file multiple times,
        # since we might have different variables set.
        -IncludeAgain => 1,

        # allow "include <path>"
        -UseApacheInclude => 1,

        # provide the ->array, ->hash, etc. methods
        -ExtendedAccess => 1,

        # interpolate config options when referred to as $foobar
        -InterPolateVars => 1,
    );

    say 'FYI: Configuration was read from the following files:';
    say '  ' . abs_path($_) for $conf->files;

    # sanity check: are there any plugins configured?
    if (!defined($conf->keys('monitor')) ||
        scalar $conf->keys('monitor') == 0) {
        say STDERR 'Your configuration does not contain any <monitor> blocks.';
        say STDERR
            'Without these blocks, running this program does not make sense.';
        exit 1;
    }

    # also: are there any jabber accounts configured?
    if (!$conf->exists('jabber')) {
        say STDERR 'Your configuration does not contain any <jabber> blocks.';
        say STDERR
            'Without these blocks, running this program does not make sense.';
        exit 1;
    }

    # Collect all jabber IDs
    # to add them to our roster
    # and allow subscription requests.
    my @all_jids;
    my $sat_global;
    if ($conf->exists('send_alerts_to')) {

        # Remember that there is a global send_alerts_to directive,
        # so that we can throw errors
        # when a module does not have its own
        # and there is no global send_alerts_to.
        $sat_global = 1;
        @all_jids = (@all_jids, split("\n", $conf->value('send_alerts_to')));
    }

    my $plugin_cfgs = $conf->obj('monitor');
    for my $name ($conf->keys('monitor')) {
        my $plugin_cfg = $plugin_cfgs->obj($name);
        if (ref $plugin_cfg eq 'ARRAY') {
            say STDERR
                "ERROR: You have two <monitor> blocks with name “$name”.";
            say STDERR
"ERROR: This is not supported. Please rename one of the blocks.";
            exit 1;
        }
        if (!$plugin_cfg->exists('send_alerts_to')) {
            if (!$sat_global) {
                say STDERR
"The <monitor $name> block is missing the send_alerts_to directive";
                exit 1;
            }
            next;
        }
        @all_jids =
            (@all_jids, split("\n", $plugin_cfg->value('send_alerts_to')));
    }

    # An AnyEvent->timer which will send @queued_messages. We need that because we
    # need to wait for presence updates to finish before we can determine an
    # inidividual user’s presence with the highest priority. While it would be
    # easier to send to a bare JID, we also need full JIDs for message receipts.
    my $queued_timer;

    $xmpp = AnyEvent::XMPP::Client->new();

    my @accounts;
    if (!$conf->is_array('jabber')) {
        @accounts = ({ $conf->hash('jabber') });
    } else {
        @accounts = $conf->array('jabber');
    }

    for my $account (@accounts) {
        $xmpp->add_account(
            $account->{jid},
            $account->{password},
            $account->{host},
            $account->{port},
            { initial_presence => undef });
    }

    my $ping = AnyEvent::XMPP::Ext::Ping->new();
    $xmpp->add_extension($ping);

    # Sends a ping request every 60 seconds. If the server does not respond within
    # another 60 seconds, reconnect.
    $ping->auto_timeout(60);

    # We are a good jabber citizen and mark this client as a bot.
    my $disco = AnyEvent::XMPP::Ext::Disco->new();
    $xmpp->add_extension($disco);
    $disco->set_identity('client', 'bot');

    # Advertise VCard support for a nice real name plus an avatar later.
    my $vcard = AnyEvent::XMPP::Ext::VCard->new();
    $disco->enable_feature($vcard->disco_feature);

    my $version = AnyEvent::XMPP::Ext::Version->new();
    $version->set_name("kanla");
    $version->set_version("0.1");
    $version->set_os("Linux");
    $xmpp->add_extension($version);
    $disco->enable_feature($version->disco_feature);

    if ($use_receipts) {
        my $receipts =
            AnyEvent::XMPP::Ext::Receipts->new(disco => $disco, debug => 1);
        $xmpp->add_extension($receipts);
    } else {
        say STDERR
"WARNING: XEP-0184 message receipts are not available because AnyEvent::XMPP is too old.";
    }

    $xmpp->set_presence(undef, 'okay (17:32:00, 2012-10-09)', 11);

    $xmpp->reg_cb(
        stream_ready => sub {
            my ($cl, $account) = @_;
            $vcard->hook_on($account->connection(), 1);
        },

        connected => sub {
            my ($self, $account) = @_;
            say "connected, adding contacts";

            # TODO: vcard avatar should be our logo as soon as we got one :)
            $vcard->store(
                $account->connection(),
                {
                    NICKNAME => 'kanla',
                    FN       => 'kanla',
                },
                sub {
                    my ($error) = @_;
                    if ($error) {
                        say "[XMPP] VCard upload failed: " . $error->string;
                    }
                });

            for my $jid (@all_jids) {
                $account->connection()->get_roster()->new_contact(
                    $jid, undef, undef,
                    sub {
                        my ($contact, $err) = @_;
                        if (defined($contact)) {
                            say "Added $jid, sending presence subscription";
                            $contact->send_subscribe();
                        } else {
                            say "Error adding $jid: $err";
                        }
                    });
            }

        },

        presence_update => sub {
            my ($cl, $account, $roster, $contact, $old_presence, $new_presence)
                = @_;

            return if defined($queued_timer);
            $queued_timer = AnyEvent->timer(

                # We wait 5 seconds for the presence updates to trickle in. On very
                # slow uplinks, that might be too short, but then again, monitoring
                # will likely not work very well anyways in that situation.
                after => 5,
                cb    => sub {
                    xmpp_empty_queue();
                    undef $queued_timer;
                });
        },

        contact_request_subscribe => sub {
            my ($cl, $acc, $roster, $contact) = @_;

            # Ignore subscription requests from people who are not in
            # @all_jids.
            return unless (scalar grep { $_ eq $contact->jid } @all_jids) > 0;

            # Acknowledge everything else.
            say "Acknowledging subscription request from " . $contact->jid;
            $contact->send_subscribed;
            $contact->send_subscribe;
        },

        disconnect => sub {
            my ($self, $account, $host, $port, $message) = @_;
            say "[XMPP] Disconnected: $message";

            # Try to reconnect, if necessary.
            $xmpp->update_connections();
        },

        error => sub {
            my ($self, $account, $error) = @_;

            # TODO: we might want to handle presence errors in a special way.
            # in case of a 404 they probably mean the user has a typo in his jid.
            say "[XMPP] Error: " . $error->string();

            # Try to reconnect, if necessary.
            $xmpp->update_connections();
        },
    );
    $xmpp->start;

    # Start all the monitoring modules,
    # read their stderr, relay errors to XMPP.
    for my $name ($conf->keys('monitor')) {
        my $plugin_cfg = $plugin_cfgs->obj($name);
        my $plugin     = $plugin_cfg->value('plugin');

        # TODO: handle send_alerts_to per plugin

        if (!defined($plugin) || $plugin eq '') {
            say STDERR
                qq|Invalid <monitor> block: 'plugin' not specified for "$name"|;
            next;
        }

        my $path = plugin_path($plugin);
        if (!-e $path) {
            say STDERR qq|Invalid <monitor> block: plugin "$plugin" not found|;
            next;
        }

        if (!-X $path) {
            say STDERR
qq|Invalid <monitor> block: plugin "$plugin" not executable (try chmod +x?)|;
            next;
        }

        start_plugin($plugin, $name);
    }
}

1

__END__

=encoding utf-8

=head1 NAME

kanla - small-scale alerting daemon

=head1 DESCRIPTION

kanla is a daemon which peridiocally checks
whether your website, mail server, etc.
are still up and running.

In case a health check fails,
kanla will notify you
via jabber (XMPP).

Focus of kanla lies on
being light-weight,
being simple,
using a sane configuration file,
being well-documented.

=head1 DOCUMENTATION

kanla's documentation can be found at
http://kanla.zekjur.net/docs/

We have decided to use asciidoc for kanla,
and to not maintain both POD and asciidoc,
the POD documentation is intentionally sparse.

=head1 VERSION

Version 1.5

=head1 AUTHOR

Michael Stapelberg, C<< <michael at stapelberg.de> >>

=head1 LICENSE AND COPYRIGHT

Copyright 2012-2014 Michael Stapelberg.

This program is free software; you can redistribute it and/or modify it
under the terms of the BSD license.

=cut

# vim:ts=4:sw=4:expandtab
